File: 2016/Level 3/level_3_part_a.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>

bool isValidTriangle(int a, int b, int c) {
    return (a + b > c) && (a + c > b) && (b + c > a);
}

int main() {
    std::ifstream inputFile("input_level_3.txt");
    if (!inputFile) {
        std::cerr << "Unable to open file input.txt" << std::endl;
        return 1;
    }

    int validTriangles = 0;
    std::string line;

    while (std::getline(inputFile, line)) {
        std::istringstream ss(line);
        int a, b, c;
        ss >> a >> b >> c;

        if (isValidTriangle(a, b, c)) {
            validTriangles++;
        }
    }

    inputFile.close();

    std::cout << "Number of valid triangles: " << validTriangles << std::endl;

    return 0;
}

File: 2024/Level 3/level_3_part_a.lua
function sum_valid_mul_results(file_path)
    local total = 0

    -- Read the entire file
    local file = io.open(file_path, "r")
    if not file then
        print("Error: Unable to open file " .. file_path)
        return nil
    end

    local data = file:read("*all")
    file:close()

    -- Pattern to match valid mul(X,Y)
    local pattern = "mul%((%d+),(%d+)%)"

    -- Iterate through all matches
    for x, y in string.gmatch(data, pattern) do
        total = total + (tonumber(x) * tonumber(y))
    end

    return total
end

-- Main
local input_file = "input_level_3.txt" -- Replace with the path to your input file
local result = sum_valid_mul_results(input_file)
if result then
    print("Sum of valid mul results: " .. result)
end